x-template.js

```diff
@@ -269,6 +270,1104 @@ class Forgiving {
     return fragment;
   }
 }
+*/
+
+// This document is a helpful summary of the spec as it relates to valid and
+//  invalid character patterns in different elements and nodes:
+//  https://w3c.github.io/html-reference/syntax.html
+//  … WHATWG also has a helpful breakdown here:
+//  https://html.spec.whatwg.org/multipage/syntax.html#elements-2
+
+// Here is a reference of all named html entities:
+//  https://html.spec.whatwg.org/multipage/named-characters.html
+//  https://html.spec.whatwg.org/entities.json
+
+// Here is a nicely-formatted list of valid HTML elements from MDN:
+//  https://developer.mozilla.org/en-US/docs/Web/HTML/Element
+//  … and a related WHATWG specification on the same:
+//  https://html.spec.whatwg.org/multipage/indices.html
+
+// TODO: Do we need to support CDATA?
+//  https://w3c.github.io/html-reference/syntax.html#cdata-sections
+//  https://developer.mozilla.org/en-US/docs/Web/API/CDATASection
+// TODO: Some of the validations around tag name, attribute name, and property
+//  name may be slow… should we be less strict? Is it the case that the browser
+//  would throw on any of these anyhow? Could we just let that happen?
+// TODO: HTML often assumed UTF-8, but JS strings (i think) are UTF-16. That
+//  means we can accept UTF-16 characters and inject them as text content, 
+// TODO: Come up with a way to test UTF-8 multi-byte sequence interop with
+//  multi-by UTF-16 surrogate pair interop. Maybe not a problem, but it would be
+//  important to know one way or the other!
+// TODO: Do we need special handling of NUL character — U+0000? It is forbidden
+//  in text / character data. For “permanently undefined Unicode characters”?
+// TODO: Can text nodes be added in svg / math namespaces?
+// TODO: Should we use setAttributeNS when in a namespace?
+// TODO: Checkout optional newline for pre / textarea:
+//  https://html.spec.whatwg.org/multipage/syntax.html#element-restrictions
+// TODO: May be more performant to turn sets / maps into switches.
+// TODO: Seems like there may be some conventions to use camelCase element
+//  names in SVG. Are we okay forcing them to be lowercase?
+
+/** Strict HTML parser meant to handle interpolated HTML. */
+class Unforgiving {
+  // More performant to clone a single fragment so we just keep one around.
+  static #fragment = new DocumentFragment();
+
+  // Special keys to hang internal state off of nodes we create for performance.
+  static #localName = Symbol();
+  static #parentNode = Symbol();
+  static #namespace = Symbol();
+
+  // Delimiter we add to improve debugging — "${…}".
+  static #delimiter = '${\u2026}';
+
+  // Self-closing html elements.
```

Comment:
TODO consider linking relevant spec stuff contextually vs in top line header?

Reply:
I went through and added much more additional commentary. I think this file is more commentary than code 😺